### ISSUE TITLE
Last update of Thor break "rails generate controller" on 4.2

### ISSUE THREAD

**Issue by benoittgt** [rails/rails#37978]
Hello

We use `rails generate` command on our `rspec-rails` builds but few days ago we saw [build failures](https://travis-ci.org/rspec/rspec-rails/builds/624990853?utm_source=github_status&utm_medium=notification) with Rails 4.2. It seems the bump of Thor dependencies is the root cause.  

### Steps to reproduce

1. `git clone https://github.com/benoittgt/rails_4_2_generate_issue`
2. `cd rails_4_2_generate_issue`
3. `bundle _1.7_ install`
4. `rails generate controller ok index`

### Expected behavior

```
$ rails generate controller ok index 
   identical  app/controllers/ok_controller.rb
       route  get 'ok/index'
      invoke  erb
      create    app/views/ok
      create    app/views/ok/index.html.erb
      invoke  test_unit
      create    test/controllers/ok_controller_test.rb
      invoke  helper
      create    app/helpers/ok_helper.rb
      invoke    test_unit
      invoke  assets
      invoke    coffee
      create      app/assets/javascripts/ok.coffee
      invoke    scss
      create      app/assets/stylesheets/ok.scss
```

### Actual behavior

```
$ rails generate controller ok index                                                                                [16:17:30]
[WARNING] Could not load generator "rails/generators/rails/controller/controller_generator". Error: undefined method `deprecation_warning' for Thor:Class
Did you mean?  deprecate_constant.
/Users/bti/.rvm/gems/ruby-2.4.6/gems/thor-1.0.0/lib/thor/parser/option.rb:140:in `validate_default_type!'
/Users/bti/.rvm/gems/ruby-2.4.6/gems/thor-1.0.0/lib/thor/parser/option.rb:115:in `validate!'
/Users/bti/.rvm/gems/ruby-2.4.6/gems/thor-1.0.0/lib/thor/parser/argument.rb:24:in `initialize'
/Users/bti/.rvm/gems/ruby-2.4.6/gems/thor-1.0.0/lib/thor/parser/option.rb:11:in `initialize'
/Users/bti/.rvm/gems/ruby-2.4.6/gems/thor-1.0.0/lib/thor/base.rb:582:in `new'
/Users/bti/.rvm/gems/ruby-2.4.6/gems/thor-1.0.0/lib/thor/base.rb:582:in `build_option'
/Users/bti/.rvm/gems/ruby-2.4.6/gems/thor-1.0.0/lib/thor/base.rb:304:in `class_option'
/Users/bti/.rvm/gems/ruby-2.4.6/gems/railties-4.2.11.1/lib/rails/generators/base.rb:202:in `class_option'
/Users/bti/.rvm/gems/ruby-2.4.6/gems/railties-4.2.11.1/lib/rails/generators/base.rb:178:in `block in hook_for'
/Users/bti/.rvm/gems/ruby-2.4.6/gems/railties-4.2.11.1/lib/rails/generators/base.rb:168:in `each'
/Users/bti/.rvm/gems/ruby-2.4.6/gems/railties-4.2.11.1/lib/rails/generators/base.rb:168:in `hook_for'
/Users/bti/.rvm/gems/ruby-2.4.6/gems/railties-4.2.11.1/lib/rails/generators/rails/controller/controller_generator.rb:21:in `<class:ControllerGenerator>'
/Users/bti/.rvm/gems/ruby-2.4.6/gems/railties-4.2.11.1/lib/rails/generators/rails/controller/controller_generator.rb:3:in `<module:Generators>'
/Users/bti/.rvm/gems/ruby-2.4.6/gems/railties-4.2.11.1/lib/rails/generators/rails/controller/controller_generator.rb:2:in `<module:Rails>'
/Users/bti/.rvm/gems/ruby-2.4.6/gems/railties-4.2.11.1/lib/rails/generators/rails/controller/controller_generator.rb:1:in `<top (required)>'
/Users/bti/.rvm/gems/ruby-2.4.6/gems/activesupport-4.2.11.1/lib/active_support/dependencies.rb:274:in `require'
/Users/bti/.rvm/gems/ruby-2.4.6/gems/activesupport-4.2.11.1/lib/active_support/dependencies.rb:274:in `block in require'
/Users/bti/.rvm/gems/ruby-2.4.6/gems/activesupport-4.2.11.1/lib/active_support/dependencies.rb:240:in `load_dependency'
/Users/bti/.rvm/gems/ruby-2.4.6/gems/activesupport-4.2.11.1/lib/active_support/dependencies.rb:274:in `require'
/Users/bti/.rvm/gems/ruby-2.4.6/gems/railties-4.2.11.1/lib/rails/generators.rb:332:in `block (2 levels) in lookup'
/Users/bti/.rvm/gems/ruby-2.4.6/gems/railties-4.2.11.1/lib/rails/generators.rb:328:in `each'
/Users/bti/.rvm/gems/ruby-2.4.6/gems/railties-4.2.11.1/lib/rails/generators.rb:328:in `block in lookup'
/Users/bti/.rvm/gems/ruby-2.4.6/gems/railties-4.2.11.1/lib/rails/generators.rb:327:in `each'
/Users/bti/.rvm/gems/ruby-2.4.6/gems/railties-4.2.11.1/lib/rails/generators.rb:327:in `lookup'
/Users/bti/.rvm/gems/ruby-2.4.6/gems/railties-4.2.11.1/lib/rails/generators.rb:138:in `find_by_namespace'
/Users/bti/.rvm/gems/ruby-2.4.6/gems/railties-4.2.11.1/lib/rails/generators.rb:155:in `invoke'
/Users/bti/.rvm/gems/ruby-2.4.6/gems/railties-4.2.11.1/lib/rails/commands/generate.rb:13:in `<top (required)>'
/Users/bti/.rvm/gems/ruby-2.4.6/gems/activesupport-4.2.11.1/lib/active_support/dependencies.rb:274:in `require'
/Users/bti/.rvm/gems/ruby-2.4.6/gems/activesupport-4.2.11.1/lib/active_support/dependencies.rb:274:in `block in require'
/Users/bti/.rvm/gems/ruby-2.4.6/gems/activesupport-4.2.11.1/lib/active_support/dependencies.rb:240:in `load_dependency'
/Users/bti/.rvm/gems/ruby-2.4.6/gems/activesupport-4.2.11.1/lib/active_support/dependencies.rb:274:in `require'
/Users/bti/.rvm/gems/ruby-2.4.6/gems/railties-4.2.11.1/lib/rails/commands/commands_tasks.rb:123:in `require_command!'
/Users/bti/.rvm/gems/ruby-2.4.6/gems/railties-4.2.11.1/lib/rails/commands/commands_tasks.rb:130:in `generate_or_destroy'
/Users/bti/.rvm/gems/ruby-2.4.6/gems/railties-4.2.11.1/lib/rails/commands/commands_tasks.rb:50:in `generate'
/Users/bti/.rvm/gems/ruby-2.4.6/gems/railties-4.2.11.1/lib/rails/commands/commands_tasks.rb:39:in `run_command!'
/Users/bti/.rvm/gems/ruby-2.4.6/gems/railties-4.2.11.1/lib/rails/commands.rb:17:in `<top (required)>'
bin/rails:4:in `require'
bin/rails:4:in `<main>'
   identical  app/controllers/ok_controller.rb
/Users/bti/.rvm/gems/ruby-2.4.6/gems/railties-4.2.11.1/lib/rails/generators/rails/controller/controller_generator.rb:16:in `block in add_routes': undefined method `generate_routing_code' for #<Rails::Generators::ControllerGenerator:0x00007fa356be67d8> (NoMethodError)
	from /Users/bti/.rvm/gems/ruby-2.4.6/gems/railties-4.2.11.1/lib/rails/generators/rails/controller/controller_generator.rb:15:in `reverse_each'
	from /Users/bti/.rvm/gems/ruby-2.4.6/gems/railties-4.2.11.1/lib/rails/generators/rails/controller/controller_generator.rb:15:in `add_routes'
	from /Users/bti/.rvm/gems/ruby-2.4.6/gems/thor-1.0.0/lib/thor/command.rb:27:in `run'
	from /Users/bti/.rvm/gems/ruby-2.4.6/gems/thor-1.0.0/lib/thor/invocation.rb:127:in `invoke_command'
	from /Users/bti/.rvm/gems/ruby-2.4.6/gems/thor-1.0.0/lib/thor/invocation.rb:134:in `block in invoke_all'
	from /Users/bti/.rvm/gems/ruby-2.4.6/gems/thor-1.0.0/lib/thor/invocation.rb:134:in `each'
	from /Users/bti/.rvm/gems/ruby-2.4.6/gems/thor-1.0.0/lib/thor/invocation.rb:134:in `map'
	from /Users/bti/.rvm/gems/ruby-2.4.6/gems/thor-1.0.0/lib/thor/invocation.rb:134:in `invoke_all'
	from /Users/bti/.rvm/gems/ruby-2.4.6/gems/thor-1.0.0/lib/thor/group.rb:232:in `dispatch'
	from /Users/bti/.rvm/gems/ruby-2.4.6/gems/thor-1.0.0/lib/thor/base.rb:476:in `start'
	from /Users/bti/.rvm/gems/ruby-2.4.6/gems/railties-4.2.11.1/lib/rails/generators.rb:157:in `invoke'
	from /Users/bti/.rvm/gems/ruby-2.4.6/gems/railties-4.2.11.1/lib/rails/commands/generate.rb:13:in `<top (required)>'
	from /Users/bti/.rvm/gems/ruby-2.4.6/gems/activesupport-4.2.11.1/lib/active_support/dependencies.rb:274:in `require'
	from /Users/bti/.rvm/gems/ruby-2.4.6/gems/activesupport-4.2.11.1/lib/active_support/dependencies.rb:274:in `block in require'
	from /Users/bti/.rvm/gems/ruby-2.4.6/gems/activesupport-4.2.11.1/lib/active_support/dependencies.rb:240:in `load_dependency'
	from /Users/bti/.rvm/gems/ruby-2.4.6/gems/activesupport-4.2.11.1/lib/active_support/dependencies.rb:274:in `require'
	from /Users/bti/.rvm/gems/ruby-2.4.6/gems/railties-4.2.11.1/lib/rails/commands/commands_tasks.rb:123:in `require_command!'
	from /Users/bti/.rvm/gems/ruby-2.4.6/gems/railties-4.2.11.1/lib/rails/commands/commands_tasks.rb:130:in `generate_or_destroy'
	from /Users/bti/.rvm/gems/ruby-2.4.6/gems/railties-4.2.11.1/lib/rails/commands/commands_tasks.rb:50:in `generate'
	from /Users/bti/.rvm/gems/ruby-2.4.6/gems/railties-4.2.11.1/lib/rails/commands/commands_tasks.rb:39:in `run_command!'
	from /Users/bti/.rvm/gems/ruby-2.4.6/gems/railties-4.2.11.1/lib/rails/commands.rb:17:in `<top (required)>'
	from bin/rails:4:in `require'
	from bin/rails:4:in `<main>'
```

### System configuration
**Rails version**: Rails 4.2.11.1

**Ruby version**: 2.4.6

### Questions

As a quick workaround, you can revert this commit https://github.com/benoittgt/rails_4_2_generate_issue/commit/355d5923b8b9ad7ae35916bb99c0ca0f1e4adc69 and stick Thor version to under `1.0` in your project's Gemfile.

I'm gonna look at why we have the error on Thor.

Cheers <3

Related: https://github.com/rspec/rspec-rails/pull/2229

**Comment by benoittgt** [comment]
I am wondering if it could be linked to.

https://github.com/rails/rails/blob/2a3f759eef10352bedce5f13b12dbdda30aacab2/railties/lib/rails/generators/app_base.rb#L355-L358

At first sight. We are properly loading Thor and not Thor bundled by bundler I just don't understand why method from `lib/thor.rb` are not referenced.

**Comment by rafaelfranca** [comment]
It should be fixed now on thor 1.0.1.
